starport/pkg/jsonfile/jsonfile.go

```diff
@@ -0,0 +1,303 @@
+	var ext bytes.Buffer
+	tarballPath, err := tarball.ExtractFile(&buf, &ext, tarballFileName)
+	if err != nil && err != tarball.ErrNotGzipType {
+		return nil, err
+	} else if err == nil {
+		// Erase the tarball bite code from the file and copy the correct one
+		if err := truncate(file, 0); err != nil {
+			return nil, err
+		}
+		if _, err := io.Copy(file, &ext); err != nil {
+			return nil, err
+		}
+	}
+
+	return &JSONFile{
+		updates:     make(map[string][]byte),
+		file:        file,
+		url:         url,
+		tarballPath: tarballPath,
+	}, nil
+}
+
+// Param return the param by key and the position into byte slice from the file reader.
+// Key can be a path to a nested parameter eg: app_state.staking.accounts
+func (g *JSONFile) Param(key string, param interface{}) (int64, error) {
+	if err := g.Reset(); err != nil {
+		return 0, err
+	}
+	dec := json.NewDecoder(g.file)
+	// Split the keys by the separator to find nested JSON parameters eg: app_state.staking.accounts
+	keys := strings.Split(key, keySeparator)
+	// Instead of unmarshal the whole content of a file
+	// this will decode one line/record at a time
+	for {
+		t, err := dec.Token()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, err
+		}
+		name, ok := t.(string)
+		if !ok {
+			continue
+		}
+		// If the nested path was found
+		if name == keys[0] {
+			if len(keys) > 1 {
+				keys = keys[1:]
+				continue
+			}
+
+			// Try to decode the all data first
+			err := dec.Decode(&param)
+			if err == nil {
+				return dec.InputOffset(), nil
+			}
+
+			// If not decode, only get the JSON value from the key
+			t, err := dec.Token()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return 0, err
+			}
+			switch t := t.(type) {
+			case int:
+				param = strconv.Itoa(t)
+			case string:
+				param = t
+			default:
+				return 0, ErrInvalidValueType
+			}
+			return dec.InputOffset(), nil
+		}
+	}
+	return 0, errors.Wrap(ErrParamNotFound, key)
+}
+
+// WithKeyValue update a file value object by key
+func WithKeyValue(key string, value string) UpdateFileOption {
+	return func(update map[string][]byte) {
+		update[key] = []byte(`"` + value + `"`)
+	}
+}
+
+// WithTime update a time value
+func WithTime(key string, t int64) UpdateFileOption {
+	return func(update map[string][]byte) {
+		formatted := time.Unix(t, 0).UTC().Format(time.RFC3339Nano)
+		update[key] = []byte(`"` + formatted + `"`)
+	}
+}
+
+// WithKeyIntValue update a file int value object by key
+func WithKeyIntValue(key string, value int) UpdateFileOption {
+	return func(update map[string][]byte) {
+		update[key] = []byte{byte(value)}
+	}
+}
+
+// Update updates the file file with the new parameters by key
+func (g *JSONFile) Update(opts ...UpdateFileOption) error {
+	for _, opt := range opts {
+		opt(g.updates)
+	}
+	if err := g.Reset(); err != nil {
+		return err
+	}
+	_, err := io.Copy(g, g.file)
+	return err
+}
+
+// Write implement the write method for io.Writer interface
```

Comment:
I think it is a deafult comment from IDE

Reply:
Sorry, I don't understand. What do you mean?

Reply:
I thought this comment was automatically generated by IDE. But it seems I was wrong.